iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.inl

```diff
@@ -0,0 +1,274 @@
+    // write posiion and read position (which causes more contention)
+
+    auto writePosition = loadNextWritePosition();
+    auto readPosition = loadNextReadPosition();
+    auto value = loadValueAt(readPosition);
+
+    auto isFull = writePosition.getIndex() == readPosition.getIndex() && readPosition.isOneCycleBehind(writePosition);
+
+    if (isFull)
+    {
+        if (tryToGainOwnershipAt(readPosition))
+        {
+            index = value.getIndex();
+            return true;
+        }
+    }
+
+    // otherwise someone else has dequeued an index and the queue was not full at the start of this popIfFull
+    return false;
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::empty()
+{
+    auto oldReadIndex = m_readPosition.load(std::memory_order_relaxed);
+    auto value = m_cells[oldReadIndex.getIndex()].load(std::memory_order_relaxed);
+
+    // if m_readPosition is ahead by one cycle compared to the value stored at head,
+    // the queue was empty at the time of the loads above (but might not be anymore!)
+    return value.isOneCycleBehind(oldReadIndex);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index IndexQueue<Capacity, ValueType>::loadNextReadPosition() const
+{
+    return m_readPosition.load(std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index IndexQueue<Capacity, ValueType>::loadNextWritePosition() const
+{
+    return m_writePosition.load(std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+typename IndexQueue<Capacity, ValueType>::Index
+IndexQueue<Capacity, ValueType>::loadValueAt(typename IndexQueue<Capacity, ValueType>::Index position) const
+{
+    return m_cells[position.getIndex()].load(std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::tryToPublishAt(typename IndexQueue<Capacity, ValueType>::Index writePosition,
+                                                     Index& oldValue,
+                                                     Index newValue)
+{
+    return m_cells[writePosition.getIndex()].compare_exchange_strong(
+        oldValue, newValue, std::memory_order_relaxed, std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+void IndexQueue<Capacity, ValueType>::updateNextWritePosition(Index& writePosition)
+{
+    // compare_exchange updates writePosition if NOT successful
+    Index newWritePosition(writePosition + 1);
+    auto updateSucceeded = m_writePosition.compare_exchange_strong(
+        writePosition, newWritePosition, std::memory_order_relaxed, std::memory_order_relaxed);
+}
+
+template <uint64_t Capacity, typename ValueType>
+bool IndexQueue<Capacity, ValueType>::tryToGainOwnershipAt(Index& oldReadPosition)
```

Comment:
Please get rid of this unnecessary abstraction, same argument as in loadNextReadPosition

Reply:
Same counter argument as well.